File: chapter9/python/safefile/safefile.py
"""
File processing functions for managed files.

readFile - Read a file
writeFile - Write a file
safeGetState - Get the recovery state for a file
safeRecover - Initiate recovery for a file
safeReadFile - Read with recovery support
safeWriteFile - Write with recovery support
"""
from os import unlink, rename
from os.path import exists, isfile, isdir

# error and message constants
NO_ERROR = 0
INVALID_NAME= 100
DOES_NOT_EXIST = 101
IS_NOT_A_FILE = 102
READ_ERROR = 103
WRITE_ERROR = 104
SAFE_NORMAL = 0
SAFE_RECOVERABLE = 110
SAFE_INTERVENE = 111

MSG_INVALID_NAME = "File name missing or not valid"
MSG_IS_NOT_A_FILE = "File {0} is not a file"
MSG_DOES_NOT_EXIST = "File {0} does not exist"
MSG_READ_ERROR = "Error reading file {0}: {1}"
MSG_WRITE_ERROR = "Error writing file {0}: {1}"

class SafeFileError (Exception):
    """
    Error definition thrown when an error occurs.
    Args
        code: Error number
        message: Text message, suitable for display
    """
    def __init__ (self, code, message):
        super (SafeFileError, self).__init__ ()
        self.code = code
        self.message = message

def readFile (file):
    """
    Read file.
    Args:
      file (str): Path / file name of file to read.
    Returns:
      data (str): Data read.
    Raises:
      SafeFileError
    """
    if file is None:
        raise SafeFileError (INVALID_NAME, MSG_INVALID_NAME)

    info = _getFileInfo (file)
    if not info["exists"]:
        raise SafeFileError (DOES_NOT_EXIST, MSG_DOES_NOT_EXIST.format (file))

    if not info["isFile"]:
        raise SafeFileError (IS_NOT_A_FILE, MSG_IS_NOT_A_FILE.format (file))

    # read data file
    try:
        data = open (file).read ()
        return data
    except IOError as e:
        raise SafeFileError (READ_ERROR,
            MSG_READ_ERROR.format (file, e.strerror))

def writeFile (file, data):
    """
    Write file.
    Args:
        file (str): Path / file name to write.
        data (str): Data to write.
    Raises:
        SafeFileError
    """
    if file is None:
        raise SafeFileError (INVALID_NAME, MSG_INVALID_NAME)

    info = _getFileInfo (file)
    if info["exists"] and not info["isFile"]:
        raise SafeFileError (IS_NOT_A_FILE, MSG_IS_NOT_A_FILE.format (file))

    # read data file
    try:
        file = open (file, "w")
        file.write (data)
        file.close ()
    except IOError as e:
        raise SafeFileError (WRITE_ERROR,
            MSG_WRITE_ERROR.format (file, e.strerror))

def safeGetState (file):
    """
    Get the status of a file in the recovery context.
    Args:
        file File to get status for.
    Retuns:
        State, can be an error or recovery state.
    """
    if file is None:
        return INVALID_NAME

    info = _getFileInfo (file)
    if info["exists"] and not info["isFile"]:
        return IS_NOT_A_FILE

    state = _getState (file)
    return state["status"]

def safeRecover (file):
    """
    Initiate the recovery processing for a file.
    Args:
        file File to performing processing for.
    Raises:
        SafeFileError
    """
    if file is None:
        raise SafeFileError (INVALID_NAME, MSG_INVALID_NAME)

    info = _getFileInfo (file)
    if info["exists"] and not info["isFile"]:
        raise SafeFileError (IS_NOT_A_FILE, MSG_IS_NOT_A_FILE.format (file))

    state = _getState (file)
    if state["status"] == DOES_NOT_EXIST:
        raise SafeFileError (DOES_NOT_EXIST, MSG_DOES_NOT_EXIST.format (file))

    _performRecovery (state, True)

def safeReadFile (file):
    """
    Read a file, applying recovery processing if necessary.
    Args:
        file File to read.
    Returns:
        Data read from file.
    Raises:
        SafeFileError
    """
    if file is None:
        raise SafeFileError (INVALID_NAME, MSG_INVALID_NAME)

    info = _getFileInfo (file)
    if info["exists"] and not info["isFile"]:
        raise SafeFileError (IS_NOT_A_FILE, MSG_IS_NOT_A_FILE.format (file))

    state = _getState (file)
    if state["status"] == SAFE_RECOVERABLE:
        _performRecovery (state, True)

    readFile (file)

def safeWriteFile (file, data):
    """
    Write data to a file, applying recovery enabling processing.
    Args:
        file File to write to.
        data Data to write.
    Raises:
        SafeFileError
    """
    if file is None:
        raise SafeFileError (INVALID_NAME, MSG_INVALID_NAME)

    info = _getFileInfo (file)
    if info["exists"] and not info["isFile"]:
        raise SafeFileError (IS_NOT_A_FILE, MSG_IS_NOT_A_FILE.format (file))

    # get current file system state, and auto-recover if necessary
    state = _getState (file)

    # store data in well defined ephemeral file to allow manual recovery.
    # If file already exists, remove it (failed prior recovery).
    if state["ephemeral"]["exists"]:
        unlink (state["ephemeral"]["name"])

    writeFile (state["ephemeral"]["name"], data)
    state["ephemeral"]["exists"] = True

    # if ready state file already exists, recover prior state
    if state["ready"]["exists"]:
        _performRecovery (state, False)

    rename (state["ephemeral"]["name"], state["ready"]["name"])

    # refresh state and process recovery to set file system state
    state = _getState (file)
    _performRecovery (state, True)

def _getState (file):
    """
    Get file state.
    Args:
        file File to get state for.
    Returns:
        State object containing list of recovery files and overall status.
    """
    state = {}
    state["ephemeral"] = _getFileInfo (file + ".eph")
    state["ready"] = _getFileInfo (file + ".rdy")
    state["base"] = _getFileInfo (file)
    state["backup"] = _getFileInfo (file + ".bak")
    state["tertiary"] = _getFileInfo (file + ".bk2")

    if state["ephemeral"]["exists"]:
        state["status"] = SAFE_INTERVENE
    elif state["ready"]["exists"] or state["tertiary"]["exists"]:
        state["status"] = SAFE_RECOVERABLE
    elif state["base"]["exists"]:
        state["status"] = SAFE_NORMAL
    else:
        if state["backup"]["exists"]:
            state["status"] = SAFE_RECOVERABLE
        else:
            state["status"] = DOES_NOT_EXIST

    return state

def _getFileInfo (file):
    """
    Get information for a file (name, exists, is a file or directory.
    Args:
        file File to get information for.
    Returns:
        Dict with file info (name, exists, isFile, isDirectory)
    """
    info = {}
    info["name"] = file
    info["exists"] = exists (file)
    info["isFile"] = isfile (file)
    info["isDirectory"] = isdir (file)
    return info

def _performRecovery (state, removeEphemeral):
    """
    Initiate recovery processing.
    Args:
        state State object with recovery file information.
        removeEphemeral Flag, remove ephemeral if found or not
    """
    # if ephemeral flag true, and ephemeral file exists, remove it
    if removeEphemeral and state["ephemeral"]["exists"]:
        unlink (state["ephemeral"]["name"])

    # if only backups exist, restore from backup
    baseAvailable = state["base"]["exists"] or state["ready"]["exists"]
    if not baseAvailable:
        if state["tertiary"]["exists"]:
            if state["backup"]["exists"]:
                rename (state["backup"]["name"], state["base"]["name"])
                rename (state["tertiary"]["name"], state["backup"]["name"])
            else:
                rename (state["tertiary"]["name"], state["base"]["name"])
        elif state["backup"]["exists"]:
            rename (state["backup"]["name"], state["base"]["name"])
        return

    # if tertiary state file exists, remove it
    if state["tertiary"]["exists"]:
        unlink (state["tertiary"]["name"])

    # if ready state file exists, update ready, base and backup files
    if state["ready"]["exists"]:
        removeTertiary = False

        # if base and backup exist, rename to tertiary temporarily
        if state["base"]["exists"] and state["backup"]["exists"]:
            rename (state["backup"]["name"], state["tertiary"]["name"])
            removeTertiary = True

        # if base exists, rename to backup
        if state["base"]["exists"]:
            rename (state["base"]["name"], state["backup"]["name"])

        # place ready state file in base and delete temporary tertiary file
        rename (state["ready"]["name"], state["base"]["name"])

        # if temporary tertiary created, remove it
        if removeTertiary:
            unlink (state["tertiary"]["name"])

File: chapter9/nodejs/safefile/safeFile.js
/**
 * File processing functions for managed files.
 * 
 * readFileSync - Read a file
 * writeFileSync - Write a file
 * safeGetState - Get the recovery state for a file
 * safeRecover - Initiate recovery for a file
 * safeReadFileSync - Read with recovery support 
 * safeWriteFileSync - Write with recovery support
 */
// import format function and Node.js file system module
var format = require ("ujs-format").format;
var fs = require ("fs");
var SafeFileError = require ("./SafeFileError").SafeFileError;
var cc = SafeFileError.prototype;

/**
 * Read file.
 * @param {String} fileName File to read.
 * @returns {String} Data read from file.
 * @throws SafeFileError
 */
function readFileSync (fileName, options) {
    verifyFileName (fileName);

    var info = getFileInfo (fileName);
    if (info.exists === false) {
        var message1 = format (cc.MSG_DOES_NOT_EXIST, fileName);
        throw new SafeFileError (cc.DOES_NOT_EXIST, message1);
    }
    if (info.isFile === false) {
        var message2 = format (cc.MSG_IS_NOT_A_FILE, fileName);
        throw new SafeFileError (cc.IS_NOT_A_FILE, message2);
    }

    // read data file
    var data = null;
    try {
        data = fs.readFileSync (fileName, options);
    } catch (e) {
        var message3 = format (cc.MSG_READ_ERROR, fileName, e.message);
        throw new SafeFileError (cc.READ_ERROR, message3);
    }

    // return data
    return (data);
}

/**
 * Write data to a file.
 * @param {String} fileName Name of file (path optional).
 * @param {String} data Data to write.
 * @throws SafeFileError
 */
function writeFileSync (fileName, data, options) {
    verifyFileName (fileName);
    
    var info = getFileInfo (fileName);
    if ((info.exists === true) && (info.isFile === false)) {
        var message1 = format (cc.MSG_IS_NOT_A_FILE, fileName);
        throw new SafeFileError (cc.IS_NOT_A_FILE, message1);
    }

    // if content undefined or null, set data to empty string
    if ((data === undefined) || (data === null)) {
        data = "";
    }

    // write file content, throwing exception on error occurring
    try {
        fs.writeFileSync (fileName, data, options);
    } catch (e) {
        var message2 = format (cc.MSG_WRITE_ERROR, e.message);
        throw new SafeFileError (cc.WRITE_ERROR, message2);
    }
}

/**
 * Get status of the file.
 * @param {String} file Name of base file.
 * @returns {Integer} SAFE_NORMAL, SAFE_AUTO_RECOVERABLE, SAFE_INTERVENE,
 *     INVALID_NAME, IS_NOT_A_FILE, or DOES_NOT_EXIST
 */
function safeGetState (fileName) {
    if ((fileName === undefined) || (fileName === null)) {
        return (cc.INVALID_NAME);
    }
    // if fileName exists, verify it is a file
    var info = getFileInfo (fileName);
    if ((info.exists) && (info.isFile === false)) {
        return (cc.IS_NOT_A_FILE);
    }

    var state = getState (fileName);
    return (state.status);
}

/**
 * Initiate auto-recovery processing.
 * @param {String} fileName Name of base file
 * @throws SafeFileError
 */
function safeRecover (fileName) {
    verifyFileName (fileName);

    // if fileName exists, verify it is a file
    var info = getFileInfo (fileName);
    if ((info.exists) && (info.isFile === false)) {
        var message1 = format (cc.MSG_IS_NOT_A_FILE, fileName);
        throw new SafeFileError (cc.IS_NOT_A_FILE, message1);
    }

    // get state, if doesn't exist throw error
    var state = getState (fileName);
    if (state.status === cc.DOES_NOT_EXIST) {
        var message2 = format (cc.MSG_DOES_NOT_EXIST, fileName);
        throw new SafeFileError (cc.DOES_NOT_EXIST, message2);
    }
        
    performRecovery (state, true);
}

/**
 * Read a file, performing recovery processing if necessary.
 * @param {String} file File to read.
 * @returns {String} Data read from file.
 * @throws SafeFileError
 */
function safeReadFileSync (fileName, options) {
    verifyFileName (fileName);
    
    // if fileName exists, verify it is a file
    var info = getFileInfo (fileName);
    if ((info.exists) && (info.isFile === false)) {
        var message = format (cc.MSG_IS_NOT_A_FILE, fileName);
        throw new SafeFileError (cc.IS_NOT_A_FILE, message);
    }

    // get state, if auto-recovery required, perform recovery
    var state = getState (fileName);
    if (state.status === cc.SAFE_RECOVERABLE) {
        performRecovery (state, true);
    }

    // perform read on file
    return (readFileSync (fileName, options));
}

/**
 * Write data to a file using a recoverable process.
 * @param {String} fileName Name of file (path optional).
 * @param {String} data Data to write.
 * @throws SafeFileError
 */
function safeWriteFileSync (fileName, data, options) {
    verifyFileName (fileName);
    
    // if fileName exists, verify it is a file
    var info = getFileInfo (fileName);
    if ((info.exists) && (info.isFile === false)) {
        var message = format (cc.MSG_IS_NOT_A_FILE, fileName);
        throw new SafeFileError (cc.IS_NOT_A_FILE, message);
    }

    // get current file system state, and auto-recover if necessary
    var state = getState (fileName);

    // store data in well defined ephemeral file to allow manual recovery
    // If file already exists, remove it (failed prior recovery).
    if (state.ephemeral.exists) {
        fs.unlinkSync (state.ephemeral.name);
    }

    writeFileSync (state.ephemeral.name, data, options);
    state.ephemeral.exists = true;

    // if ready state file already exists, recover prior state
    if (state.ready.exists) {
        performRecovery (state, false);
    }
    
    fs.renameSync (state.ephemeral.name, state.ready.name);

    // refresh state and process recovery to set file system state
    state = getState (fileName);
    performRecovery (state, true);
}

/**
 * Verify fileName parameter.
 * @param {String} fileName Name of file to verify.
 * @throws SafeFileError
 */
function verifyFileName (fileName)
{
    // if fileName undefined or null, throw exception
    if ((fileName === undefined) || (fileName === null)) {
        var message1 = format (cc.MSG_INVALID_NAME);
        throw new SafeFileError (cc.INVALID_NAME, message1);
    }
}

/**
 * Get state for file system entities.
 * @param {String} file Base file name.
 * @returns {Object} State object.
 */
function getState (file) {
    // collect state for all possible data and recovery files
    var state = {};
    state.ephemeral = getFileInfo (file + ".eph");
    state.ready = getFileInfo (file + ".rdy");
    state.base = getFileInfo (file);
    state.backup = getFileInfo (file + ".bak");
    state.tertiary = getFileInfo (file + ".bk2");

    if (state.ephemeral.exists) {
        state.status = cc.SAFE_INTERVENE;
    } else if ((state.ready.exists) || (state.tertiary.exists)) {
        state.status = cc.SAFE_RECOVERABLE;
    } else if (state.base.exists) {
        state.status = cc.SAFE_NORMAL;
    } else {
        if (state.backup.exists) {
            state.status = cc.SAFE_RECOVERABLE;
        } else {
            state.status = cc.DOES_NOT_EXIST;
        }
    }

    return (state);
}

/**
 * Get existence, file/directory info for a file.
 * @param {String} fileName Name of file to get info for
 * @returns {Object}
 */
function getFileInfo (fileName) {
    var result = {};
    result.name = fileName;
    result.exists = fs.existsSync (fileName);
    if (result.exists) {
        var stats = fs.statSync (fileName);
        result.isFile = stats.isFile ();
        result.isDirectory = stats.isDirectory ();
    }
    else {
        result.isFile = false;
        result.isDirectory = false;
    }
    
    return (result);
}

/**
 * Evaluate save state, initiating recovery if necessary.
 * @param {Object} state State object with file names and existence flags
 */
function performRecovery (state, removeEphemeral) {
    // if ephemeral flag true, and ephemeral file exists, remove it
    if ((removeEphemeral) && (state.ephemeral.exists)) {
        fs.unlinkSync (state.ephemeral.name);
    }

    // if only backups exist, restore from backup
    var baseAvailable = state.base.exists || state.ready.exists;
    if (baseAvailable === false) {
        if (state.tertiary.exists) {
            if (state.backup.exists) {
                fs.renameSync (state.backup.name, state.base.name);
                fs.renameSync (state.tertiary.name, state.backup.name);
            } else {
                fs.renameSync (state.tertiary.name, state.base.name);
            }
        } else if (state.backup.exists) {
            fs.renameSync (state.backup.name, state.base.name);
        }
        
        return;
    }

    // if tertiary state file exists, remove it
    if (state.tertiary.exists) {
        fs.unlinkSync (state.tertiary.name);
    }

    // if ready state file exists, update ready, base and backup files
    if (state.ready.exists) {
        var removeTertiary = false;
        

        // if base and backup exist, rename to tertiary temporarily
        if ((state.base.exists) && (state.backup.exists)) {
            fs.renameSync (state.backup.name, state.tertiary.name);
            removeTertiary = true;
        }

        // if base exists, rename to backup
        if (state.base.exists) {
            fs.renameSync (state.base.name, state.backup.name);
        }

        // place ready state file in base and delete temporary tertiary file
        fs.renameSync (state.ready.name, state.base.name);

        // if temporary tertiary created, remove it
        if (removeTertiary) {
            fs.unlinkSync (state.tertiary.name);
        }
    }
}

// exports
exports.readFileSync = readFileSync;
exports.writeFileSync = writeFileSync;
exports.safeGetState = safeGetState;
exports.safeRecover = safeRecover;
exports.safeReadFileSync = safeReadFileSync;
exports.safeWriteFileSync = safeWriteFileSync;
